### PR TITLE
warn about outdated Flutter installations

### LINKED ISSUE #6251: tool should warn when users are out of date
If the flutter checkout that people are using is more than a week old, if there is a newer checkin on the selected branch, we should display a message whenever the tool starts up saying so. Maybe only do this once per day or something.

```
Your flutter installation is 3 weeks old. To update to the latest version, run flutter update.
```

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes #6251

This banner will appear when the user runs any `flutter` command. It will fetch from GitHub no more frequently than once per week.

Sample warning message (it won't actually warn about a 4-day-old installation; I forced it to print for demo purposes):

example-warning

/cc @tvolkert @Hixie

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/clock.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// found in the LICENSE file.
+
+import 'package:quiver/time.dart';
+
+import 'context.dart';
+
+/// Currently active clock implementation.
+///
+/// By default uses system clock. Override this in tests using [Clock.fixed].
+Clock get clock => context == null ? const Clock() : context[Clock];
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/clock.dart]: Instead of `context[Clock]`, you could use `context.putIfAbsent(Clock, () => const Clock())` to alleviate all the entry points from having to seed the value in the context if all they want is the default.

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -74,6 +74,16 @@ class Cache {
     _lock = null;
   }
 
+  /// Checks if the current process owns the lock for the cache directory at
+  /// this very moment; throws a [StateError] if it doesn't.
+  static void checkLockAcquired() {
+    if (!_lockEnabled && _lock == null) {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/cache.dart]: `s/!//` - if the caller has explicitly disabled locking, we shouldn't error.

File: packages/flutter_tools/lib/src/version.dart [hunk 3]
```diff
@@ -72,11 +81,59 @@ class FlutterVersion {
   }
 
   /// A date String describing the last framework commit.
-  static String get frameworkCommitDate {
-    return _runSync(<String>['git', 'log', '-n', '1', '--pretty=format:%ad', '--date=format:%Y-%m-%d %H:%M:%S'], Cache.flutterRoot);
+  static String get frameworkCommitDate => _latestGitCommitDate();
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/version.dart]: Is there any reason for these getters to remain static?

File: packages/flutter_tools/lib/src/version.dart [hunk 4]
```diff
@@ -72,11 +81,59 @@ class FlutterVersion {
...
+  /// In the absence of bugs and crashes a Flutter developer should never see
+  /// this remote appear in their `git remote` list, but also if it happens to
+  /// persist we do the proper clean-up for extra robustness.
+  static const String _kVersionCheckRemote = '__flutter_version_check__';
+
+  /// The date of the latest framework commit in the remote repository.
+  ///
+  /// Throws [ToolExit] if a git command fails, for example, when the remote git
+  /// repository is not reachable due to a network issue.
+  static String get remoteFrameworkCommitDate {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/version.dart]: We'd like to get away from using `runSync()` (#7102), so for new API, we should definitely make it return a Future and using `run()`

File: packages/flutter_tools/lib/src/version.dart [hunk 5]
```diff
@@ -96,11 +153,120 @@ class FlutterVersion {
...
+  static String versionOutOfDateMessage(Duration frameworkAge) {
+    String warning = 'WARNING: your installation of Flutter is ${frameworkAge.inDays} days old.';
+    // Append enough spaces to match the message box width.
+    warning += ' ' * (74 - warning.length);
+
+    return '''
+  ╔════════════════════════════════════════════════════════════════════════════╗
+  ║ $warning ║
+  ║                                                                            ║
+  ║ To update to the latest version, run flutter update.                       ║
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/version.dart]: `flutter upgrade`

File: packages/flutter_tools/pubspec.yaml [hunk 6]
```diff
@@ -23,6 +24,7 @@ dependencies:
   package_config: '>=0.1.5 <2.0.0'
   platform: 1.1.1
   process: 2.0.1
+  quiver: ^0.24.0
```
Comment (tvolkert) [inline packages/flutter_tools/pubspec.yaml]: why not `0.25.0`? (it's the version that exists internally in Google)

File: packages/flutter_tools/lib/src/base/clock.dart (hunk 1 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: Done

File: packages/flutter_tools/lib/src/cache.dart (hunk 2 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/cache.dart]: Done

File: packages/flutter_tools/lib/src/version.dart (hunk 3 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Doesn't look like it. Made it an instance getter.

File: packages/flutter_tools/lib/src/version.dart (hunk 4 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 5 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/pubspec.yaml (hunk 6 above)
Comment (yjbanov) [inline packages/flutter_tools/pubspec.yaml]: Because we require identical versions across all of our packages, and updating quiver is out of scope for this PR. I can send a follow-up PR upgrading quiver.

File: packages/flutter_tools/test/src/runner/flutter_command_runner_test.dart [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import 'flutter_command_test.dart';
+
+void main() {
+  group('FlutterCommandRunner', () {
+    testUsingContext('checks that Flutter installation is up-to-date', () async {
+      final MockFlutterVersion version = FlutterVersion.instance;
+      bool versionChecked = false;
+      when(version.checkFlutterVersionFreshness()).thenAnswer((_) async {
+        versionChecked = true;
+        return null;
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/runner/flutter_command_runner_test.dart]: I think you can remove this line

File: packages/flutter_tools/test/src/version_test.dart [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+void fakeData({
+  @required DateTime localCommitDate,
+  DateTime remoteCommitDate,
+  String versionCheckStamp,
+  bool expectSetStamp: false,
+  bool errorOnFetch: false,
+}) {
+  final MockProcessManager pm = context.getVariable(ProcessManager);
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/version_test.dart]: Nit: `context[ProcessManager]`

(and below)

File: packages/flutter_tools/lib/src/version.dart [hunk 9]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+  }
+
+  return '';
+}
+
+/// Runs [command] and returns the standard output as a string.
+///
+/// If [lenient] is `true` and the command fails, returns an empty string.
+/// Otherwise, throws a [ToolExit] exception.
+Future<String> _run(List<String> command, String cwd, {bool lenient: true}) async {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/version.dart]: It looks like you can remove the `cwd` argument altogether (since it has no bearing on the commands you're running) and remove the `lenient` argument and hard-code the `false` code path (since you're always passing `false`).

(same goes with the `_runSync()` method)

**Comment by tvolkert** [comment]
LGTM w/comments

File: packages/flutter_tools/lib/src/base/clock.dart [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import 'package:quiver/time.dart';
+
+import 'context.dart';
+
+const Clock _systemClock = const Clock();
+
+/// Currently active clock implementation.
+///
+/// By default uses system clock. Override this in tests using [Clock.fixed].
+Clock get clock => context == null
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/clock.dart]: nit: please don't use => for multiline expressions

File: packages/flutter_tools/lib/src/version.dart [hunk 11]
```diff
@@ -72,11 +81,59 @@ class FlutterVersion {
...
 
   /// A date String describing the last framework commit.
-  static String get frameworkCommitDate {
-    return _runSync(<String>['git', 'log', '-n', '1', '--pretty=format:%ad', '--date=format:%Y-%m-%d %H:%M:%S'], Cache.flutterRoot);
+  String get frameworkCommitDate => _latestGitCommitDate();
+
+  static String _latestGitCommitDate([String branch]) {
+    final List<String> args = <String>['git', 'log'];
+
+    if (branch != null) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nit: no braces

File: packages/flutter_tools/lib/src/version.dart [hunk 12]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+
+    // Cache is empty or it's been a while since the last server ping. Ping the server.
+    try {
+      final DateTime remoteFrameworkCommitDate = DateTime.parse(await FlutterVersion.fetchRemoteFrameworkCommitDate());
+      Cache.instance.setStampFor(kFlutterVersionCheckStampFile, kPrettyJsonEncoder.convert(<String, String>{
+        'lastTimeVersionWasChecked': '${clock.now()}',
+        'lastKnownRemoteVersion': '$remoteFrameworkCommitDate',
+      }));
+      return remoteFrameworkCommitDate;
+    } on ToolExit catch(error) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nit: space after "catch"

File: packages/flutter_tools/lib/src/version.dart [hunk 13]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+    const JsonEncoder kPrettyJsonEncoder = const JsonEncoder.withIndent('  ');
+    final String versionCheckStamp = Cache.instance.getStampFor(kFlutterVersionCheckStampFile);
+
+    if (versionCheckStamp != null) {
+      final Map<String, String> data = JSON.decode(versionCheckStamp);
+      final DateTime lastTimeVersionWasChecked = DateTime.parse(data['lastTimeVersionWasChecked']);
+      final Duration timeSinceLastCheck = clock.now().difference(lastTimeVersionWasChecked);
+
+      // Don't ping the server too often. Return cached value if it's fresh.
+      if (timeSinceLastCheck < kCheckAgeConsideredUpToDate) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nit: no braces

File: packages/flutter_tools/lib/src/version.dart [hunk 14]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+  static String versionOutOfDateMessage(Duration frameworkAge) {
+    String warning = 'WARNING: your installation of Flutter is ${frameworkAge.inDays} days old.';
+    // Append enough spaces to match the message box width.
+    warning += ' ' * (74 - warning.length);
+
+    return '''
+  ╔════════════════════════════════════════════════════════════════════════════╗
+  ║ $warning ║
+  ║                                                                            ║
+  ║ To update to the latest version, run flutter upgrade.                      ║
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: maybe `run "flutter upgrade".` to make it clear the `.` isn't part of the command?

File: packages/flutter_tools/lib/src/version.dart [hunk 15]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+    Future<bool> newerFrameworkVersionAvailable() async {
+      final DateTime latestFlutterCommitDate = await _getLatestAvailableFlutterVersion();
+
+      if (latestFlutterCommitDate == null)
+        return false;
+
+      return latestFlutterCommitDate.isAfter(localFrameworkCommitDate);
+    }
+
+    if (installationSeemsOutdated && await newerFrameworkVersionAvailable()) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nit: no braces

File: packages/flutter_tools/lib/src/version.dart [hunk 16]
```diff
@@ -72,11 +81,59 @@ class FlutterVersion {
...
+      await _removeVersionCheckRemoteIfExists();
+    }
+  }
+
+  static Future<Null> _removeVersionCheckRemoteIfExists() async {
+    final List<String> remotes = (await _run(<String>['git', 'remote'], Cache.flutterRoot, lenient: false))
+        .split('\n')
+        .map((String name) => name.trim())  // to account for OS-specific line-breaks
+        .toList();
+    if (remotes.contains(_kVersionCheckRemote)) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nit: no braces

File: packages/flutter_tools/lib/src/version.dart [hunk 17]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+
+  /// We warn the user if the age of their Flutter installation is greater than
+  /// this duration.
+  @visibleForTesting
+  static const Duration kVersionAgeConsideredUpToDate = const Duration(days: 4 * 7);
+
+  /// The amount of time we wait before pinging the server to check for the
+  /// availability of a newer version of Flutter.
+  @visibleForTesting
+  static const Duration kCheckAgeConsideredUpToDate = const Duration(days: 1);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: let's make this a week or so

Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: maybe for clarity, make kVersionAgeConsideredUpToDate explicitly something like 4 times the kCheckAgeConsideredUpToDate, so that the max error is the same regardless of what we decide the right threshold is later? Or the other way around, make kCheckAgeConsideredUpToDate a fraction of kVersionAgeConsideredUpToDate?

File: packages/flutter_tools/lib/src/version.dart [hunk 18]
```diff
@@ -96,11 +153,141 @@ class FlutterVersion {
...
+/// Otherwise, throws a [ToolExit] exception.
+String _runSync(List<String> command, String cwd, {bool lenient: true}) {
   final ProcessResult results = processManager.runSync(command, workingDirectory: cwd);
-  return results.exitCode == 0 ? results.stdout.trim() : '';
+
+  if (results.exitCode == 0)
+    return results.stdout.trim();
+
+  if (!lenient) {
+    throwToolExit(
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: ToolExit seems like a violent exception to throw here. It's also confusing because that has the semantics of "the tool will now exit" but then you catch it, which defeats those semantics. I suggest having a separate exception, e.g. "VersionCheckError" or some such, specifically for this purpose.

File: packages/flutter_tools/test/src/version_test.dart [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+  };
+
+  when(pm.runSync(any, workingDirectory: any)).thenAnswer(syncAnswer);
+  when(pm.run(any, workingDirectory: any)).thenAnswer((Invocation invocation) async {
+    return syncAnswer(invocation);
+  });
+}
+
+class MockProcessManager extends Mock implements ProcessManager {}
+class MockCache extends Mock implements Cache {}
```
Comment (Hixie) [inline packages/flutter_tools/test/src/version_test.dart]: +1 for extensive testing.

File: packages/flutter_tools/lib/src/base/clock.dart [hunk 20]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:quiver/time.dart';
+
+import 'context.dart';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/clock.dart]: this is out of scope of this PR, and I defer to @johnmccutchan in general now about this since he owns the tool, but, I gotta say, I hope we can move away from the "context" approach. It's just fancy global variables. I'd much rather see explicit injection of dependencies, e.g. if FlutterVersion took a `clock` argument and defaulted to `const Clock()`.

Comment (Hixie) [inline packages/flutter_tools/lib/src/base/clock.dart]: I'd probably also restrict the usage of quiver to the actual tests and just actually pass in a callback to get the current time but that's even less of a real problem.

File: packages/flutter_tools/test/src/runner/flutter_command_runner_test.dart (hunk 7 above)
Comment (yjbanov) [inline packages/flutter_tools/test/src/runner/flutter_command_runner_test.dart]: Done.

File: packages/flutter_tools/test/src/version_test.dart (hunk 8 above)
Comment (yjbanov) [inline packages/flutter_tools/test/src/version_test.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 9 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done 80%. `cwd` is gone from both methods. `lenient` is gone from `_run`, but I kept it in `_runAsync` because we don't always pass `false`.

File: packages/flutter_tools/lib/src/base/clock.dart (hunk 10 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 11 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 12 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 13 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 14 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 15 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 16 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/lib/src/version.dart (hunk 17 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done. 1 week for `kCheckAgeConsideredUpToDate` and `kVersionAgeConsideredUpToDate = kCheckAgeConsideredUpToDate * 4`.

File: packages/flutter_tools/lib/src/version.dart (hunk 18 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done (added `VersionCheckError`)

File: packages/flutter_tools/test/src/version_test.dart (hunk 19 above)
Comment (yjbanov) [inline packages/flutter_tools/test/src/version_test.dart]: Thanks!

File: packages/flutter_tools/lib/src/base/clock.dart (hunk 20 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: Done. Removed `Clock` from `AppContext`. Personally I have not experienced issues with the `AppContext` approach. Like with many dependency injection frameworks it has the disadvantage of making dependencies implicit. But in return it reduces some boilerplate. It's a trade-off I'm used to.

Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: Also, `AppContext` works in static and top-level functions.

Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/clock.dart]: Yeah, I'd rather not move away from AppContext until we can move to a proper DI framework, like `package:inject`.  Manual injection via constructor args always devolves into huge bloat of singletons being passed down the entire object graph from `main()` on down.

Not saying that I love `AppContext`, but I don't want the cure to be worse than the disease.

Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/clock.dart]: > Also, AppContext works in static and top-level functions.

Well... I'd like to get away from statics and top-level functions :-)

Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: > I'd like to get away from statics and top-level functions :-)

Why?

Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/clock.dart]: They're generally less test-friendly.

Comment (yjbanov) [inline packages/flutter_tools/lib/src/base/clock.dart]: > They're generally less test-friendly.

Github comments are probably not a good venue for this discussion, but I'd be curious to know what makes them less test-friendly than anything else. Something tells me we're making different assumptions :)
